mapeditor/mainwindow.cpp

```diff
@@ -429,6 +446,87 @@ void MainWindow::on_actionOpen_triggered()
 	openMap(filenameSelect);
 }
 
+void MainWindow::on_actionOpenRecent_triggered()
+{
+	QSettings s(Ui::teamName, Ui::appName);
+	QStringList recentFiles = s.value(recentlyOpenedFilesSetting).toStringList();
+
+	struct RecentFileDialog : public QDialog {
+		
+		RecentFileDialog(QWidget *parent, const QStringList& recentFiles) : QDialog(parent), layout(this) {
+
+			setWindowTitle("Recently Opened Files");
+			setMinimumWidth(600);
+
+			for (const QString &file : recentFiles) {
+				QListWidgetItem *item = new QListWidgetItem(file);
+				listWidget.addItem(item);
+			}
+
+			// Select most recent items by default.
+			// This enables a "CTRL+R => Enter"-workflow instead of "CTRL+R => 'mouse click on first item'"
+			if(listWidget.count() > 0) {
+				listWidget.item(0)->setSelected(true);
+			}
+
+			layout.setSizeConstraint(QLayout::SetMaximumSize);
+			layout.addWidget(&listWidget);
+		}
+
+		QVBoxLayout layout;
+		QListWidget listWidget;
+	};
+
+	RecentFileDialog d(this, recentFiles);
+
+	auto onSelect = [this, &d](QListWidgetItem *item) {
+		if(getAnswerAboutUnsavedChanges()) {
+			QString filename = item->text();
+			openMap(filename);
+		}
+		d.close();
+	};
+
+	connect(&d.listWidget, &QListWidget::itemActivated, this, onSelect);
+
+	d.exec();
+}
+
+void MainWindow::on_menuOpenRecent_aboutToShow()
+{
+	// Clear all actions except "More...", lest the list will grow with each
+	// showing of the list
+	for (QAction* action : ui->menuOpenRecent->actions()) {
+		if (action != ui->actionOpenRecentMore) {
+			ui->menuOpenRecent->removeAction(action);
+		}
+	}
+
+	QSettings s(Ui::teamName, Ui::appName);
+	QStringList recentFiles = s.value(recentlyOpenedFilesSetting).toStringList();
+
+	// Dynamically populate menuOpenRecent with one action per file.
+	for (const QString & file : recentFiles) {
+		QAction *action = new QAction(file, this);
+		ui->menuOpenRecent->insertAction(ui->actionOpenRecentMore, action);
+		connect(action, &QAction::triggered, this, [this, file]() {
+			if(!getAnswerAboutUnsavedChanges())
+				return;
+			openMap(file);
+		});
+	}
+
+	// Finally add a separator between recent entries and "More..."
+	if(recentFiles.size() > 0) {
+		ui->menuOpenRecent->insertSeparator(ui->actionOpenRecentMore);
+	}
+}
+
+void MainWindow::on_actionOpenRecentMore_triggered()
+{
+	on_actionOpenRecent_triggered();
+}
```

Comment:
just connect to the called slot directly

Reply:
Fixed with [d1e4041](https://github.com/vcmi/vcmi/pull/4939/commits/d1e40411dda31c6aa73bae84fefd81fb6b28920c)